pkg/web/apps/apps_psql_repository.go

```diff
@@ -128,5 +130,154 @@ func (a *appsPostgreSQLRepository) GetAppByID(ID string) (*models.App, error) {
+
+	version.DisabledMessage = disabledMessage.String
+	version.LastLaunchedAt = lastLaunchedAt.String
+
+	if err != nil {
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &version, nil
+}
+
+// GetDeviceByDeviceIDAndAppID returns a device by its device ID and app ID
+func (a *appsPostgreSQLRepository) GetDeviceByDeviceIDAndAppID(deviceID string, appID string) (*models.Device, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.device_id = $1 AND d.app_id = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, deviceID, appID).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
+
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetDeviceByVersionAndAppID returns a device by its version number and app ID
+func (a *appsPostgreSQLRepository) GetDeviceByVersionAndAppID(version string, appID string) (*models.Device, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.app_id = $1 AND d.device_version = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, appID, version).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
+
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetAppByID retrieves an app by id from the database
+func (a *appsPostgreSQLRepository) GetAppByAppID(appID string) (*models.App, error) {
+	app := models.App{}
+
+	sqlStatment := `SELECT id,app_id,app_name FROM app WHERE app_id=$1 AND deleted_at IS NULL;`
+	err := a.db.QueryRow(sqlStatment, appID).Scan(&app.ID, &app.AppID, &app.AppName)
+
+	if err != nil {
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &app, nil
+
+}
+
+// InsertVersionOrUpdateNumOfAppLaunches creates a new version row
+// or increments the num_of_app_launches counter if the version already exists
+func (a *appsPostgreSQLRepository) InsertVersionOrUpdateNumOfAppLaunches(version *models.Version) error {
+	sqlStatement := `
+		INSERT INTO version(id, version, app_id, disabled, disabled_message, num_of_app_launches, last_launched_at)
+		VALUES($1, $2, $3, $4, $5, $6, NOW())
+		ON CONFLICT (id)
+		DO UPDATE
+		SET num_of_app_launches = $6,
+		last_launched_at = NOW();`
```

Comment:
According to the docs, this uses the current local time[1]. It might be good to make sure this is always UTC in my opinion.

[1] https://golang.org/pkg/time/#Now

Reply:
@davidffrench This is the PostgreSQL `NOW()` function.
http://www.postgresqltutorial.com/postgresql-now/

Actually now that I think about it - we have two TIMESTAMP columns - `deleted_at` and `last_launched_at`. Neither store a timestamp **without time zone**.

Is this what we want? From PostgreSQL [documentation](https://www.postgresql.org/docs/9.1/datatype-datetime.html):

> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's timezone parameter, and is converted to UTC using the offset for the timezone zone.

How will the current implementation work in practice? Would it be better to store time zone information as it defaults to UTC? This probably requires a separate discussion/investigation - but its still relevant enough to be initiated here.

![ERD](https://user-images.githubusercontent.com/1596014/53108499-0d09ec80-352f-11e9-8e7c-1943e1984986.png)

Reply:
Good information. If we are always defaulting to the server time or UTC, there is no need to store the timezone. For now, I believe we can leave the fields as `TIMESTAMP WITHOUT TIMEZONE` and just use `NOW()` to store the server time. This will always be consistent.

Reply:
Also, you can do it by the code using the time.now() function. See [here](https://github.com/aerogear/mobile-security-service/blob/master/pkg/web/apps/apps_psql_repository.go#L166)